### DISCUSSION TITLE
imx477 (RaspPi HQ Camera) Capture Speed at Night

### DISCUSSION THREAD

**Discussion by chrisramsay** [aaronwmorris/indi-allsky#1592] [General]
My setup consists of a RPi 5 and a imx477 camera. I've found that the overnight captures are working at about ~15 frames per hour; each image takes about 4 minutes or so to capture and process. The image capture time I have set to 60s and the gain is set at a maximum of 22 (I believe set by the camera itself). Ar this rate I'll never get star trails because I'm unlikely to hit the required number of frames for that in a single night.

Is there anything I can do to improve captures per hour? I feel that perhaps my camera settings are sub optimal. Or maybe I need a better camera more suited to astrophotography. One thing I have considered is removing the IR filter from the current camera to improve sensitivity, but would that be enough?

**Comment by aaronwmorris** [comment]
It sounds like the Auto-White Balance feature of libcamera is biting you.  This is a documented "feature" of libcamera, but you have to dig pretty far down to get good information about it.

The AWB feature of libcamera requires the camera to take 3-5 images of a given exposure to properly determine the white balance.  This would mean a 60s exposure takes 3-5 minutes to complete.  You do not notice the issue during the daytime, but there is no noticeable lag when five 0.0003s exposures are taken.

Currently, the only fix for this is to disable libcamera's AWB algorithm.

**Reply by chrisramsay** [reply]
Okay thank you for the explanation. Would that be possible via the `--immediate` option?

**Reply by aaronwmorris** [reply]
indi-allsky already uses the --immediate option as a default parameter.  If you look at the Config page on the Camera tab, nearer to the bottom where the libcamera options are located, there are two settings `Night Enable AWB` and `Day Enable AWB`.  You can just disable the AWB option for the night time.

![image](https://github.com/user-attachments/assets/dd257caa-4921-4b42-a00a-e539d838200f)

**Reply by chrisramsay** [reply]
@aaronwmorris thanks for the pointers. As it is, both of the enable AWB options were already set to 'Disabled' and the logging does indeed indicate that the `--immediate` option is already being passed. Despite this, the capture is still slow. One other thing I have thought of is that I am using a camera cable extension that runs the camera signal via ethernet - could this be to blame? Whatever the issue, I will do some testing by using the same capture parameters as seen in the logs and see if I can figure out what could be causing the slowness.

**Reply by aaronwmorris** [reply]
If you could look through the log for the night time to see how long the exposures take to complete.  The lines in the log will look something like this:

```
Oct 28 09:08:47 allskyasi385 [INFO] Capture-1-2545/MainThread capture.saferun() [464]: Exposure received in 1.3625 s (1.3622)
```

You can do this to get a bunch of them.

```
grep "Exposure received in" /var/log/indi-allsky/indi-allsky.log
```

**Comment by chrisramsay** [comment]
This is the camera command issued:

`2024-10-28T20:01:16.769718+00:00 allsky [INFO] Capture-12-207305/MainThread libcamera.setCcdExposure() [263]: image command: rpicam-still --immediate --nopreview --camera 0 --raw --denoise off --gain 22 --shutter 30000000 --metadata /tmp/tmpdgrlj01q.json --metadata-format json --awbgains 1,1 --output /tmp/tmpybaffhgs.dng`

With a corresponding output:

`2024-10-28T20:03:18.038887+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2702 s (91.2702)`

This is a bunch from slightly earlier:

`2024-10-28T18:30:18.581674+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 91.2860 s (61.2860)`
`2024-10-28T18:32:19.848293+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2671 s (91.2671)`
`2024-10-28T18:34:21.121486+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2732 s (91.2732)`
`2024-10-28T18:36:22.401903+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2805 s (91.2805)`
`2024-10-28T18:38:23.668383+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2664 s (91.2664)`
`2024-10-28T18:40:24.937364+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2691 s (91.2691)`
`2024-10-28T18:42:26.257080+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3199 s (91.3199)`
`2024-10-28T18:44:27.524185+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2669 s (91.2669)`
`2024-10-28T18:46:28.793327+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2692 s (91.2692)`
`2024-10-28T18:48:30.111661+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3183 s (91.3183)`
`2024-10-28T18:50:31.432989+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3213 s (91.3213)`
`2024-10-28T18:52:32.750519+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3175 s (91.3175)`
`2024-10-28T18:54:34.022827+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2724 s (91.2724)`
`2024-10-28T18:56:35.286652+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.2638 s (91.2638)`
`2024-10-28T18:58:36.608372+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3217 s (91.3217)`
`2024-10-28T19:00:37.933108+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3247 s (91.3247)`
`2024-10-28T19:02:39.257410+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3243 s (91.3243)`
`2024-10-28T19:04:40.573414+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 121.3160 s (91.3160)`
`2024-10-28T19:07:11.879852+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 151.3064 s (121.3064)`
`2024-10-28T19:08:43.160350+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 91.2805 s (61.2805)`
`2024-10-28T19:10:14.447325+00:00 allsky [INFO] Capture-12-207305/MainThread capture.saferun() [427]: Exposure received in 91.2870 s (61.2870)`

**Reply by aaronwmorris** [reply]
Wow, that is quite a lag.  Can you post your Support Info text here, as well?

**Comment by chrisramsay** [comment]
Sorry that it has taken a little while to get back to you.

Please see support script output:

[support.txt](https://github.com/user-attachments/files/17683924/support.txt)

**Reply by aaronwmorris** [reply]
I have never heard of a camera behaving this way.  The long CSI cable run via ethernet is a rather unique setup, so your guess that this may have something to do with the problem is a good guess.

Do the daytime images have a similar lag?

You could try switching to JPEG mode instead of DNG, this would significantly reduce the size of the file downloaded from the camera.
